File: sim/src/core/geology.ts
/**
 * Strata / geology model (v0.8 Underground — sim side).
 *
 * Design goals, mirroring weather.ts:
 *  1. DETERMINISM. The subsurface at any column (x, y) is a *pure function* of
 *     (seed, x, y, city geology profile). There is NO stored 3D grid: band
 *     depths are reconstructed on demand from seeded value-noise, so the model
 *     costs O(1) memory, adds zero bytes to saves, and reproduces bit-for-bit.
 *     Like weather it draws from a DERIVED seed stream (seed ⊕ geology salts),
 *     never from `state.rngState`, so enabling geology cannot perturb the
 *     existing city-event / growth RNG — old replays still reproduce.
 *  2. PER-CITY PROFILES. Each city id carries a compact geology profile
 *     (nominal layer thicknesses, rock hardness, water-table behaviour),
 *     hardcoded here like the climate tables — no content rebake. Manhattan is
 *     shallow hard schist; Boston is deep fill+clay; Chicago is deep soft clay;
 *     SF is mixed with bay mud; and so on.
 *  3. WATER TABLE from elevation. The table is shallow in low ground near water
 *     and deeper on high ground, computed from the city's StaticElevation
 *     surface height relative to sea level plus the profile's wetness.
 *
 * The tunable *economics* (what a tunnel costs to bore through each stratum)
 * live in the sibling module `geologyCost.ts`, exactly as weatherEffects.ts
 * holds the weather economics — one file for balance passes.
 */
import type { Vec2 } from './geometry';

/** Top-down stratum kinds. `fill` = made ground / soil, `clay` = clay/sand
 *  mixed overburden, `rock` = competent rock, `bedrock` = deep basement rock. */
export type Stratum = 'fill' | 'clay' | 'rock' | 'bedrock';

/** Canonical top-down order. */
export const STRATA: readonly Stratum[] = ['fill', 'clay', 'rock', 'bedrock'] as const;

/** A single band in a reconstructed column. Depths are metres BELOW the
 *  surface (top < bottom); the bedrock band uses a large nominal bottom. */
export interface StrataBand {
  kind: Stratum;
  top: number;
  bottom: number;
}

/** A fully reconstructed subsurface column at one (x, y). */
export interface StrataColumn {
  bands: StrataBand[];
  /** depth (m below surface) to the top of the water table */
  waterTableDepth: number;
  /** 0..1 — how hard the competent rock is to cut (schist/granite high) */
  rockHardness: number;
  /** surface elevation (m above sea level) used to build this column */
  surfaceElevation: number;
}

/** Nominal bottom of the (unbounded) bedrock band, metres. */
export const BEDROCK_NOMINAL_BOTTOM = 1000;

// ── Per-city geology profiles (content, like the climate tables) ─────────────
// Thicknesses are nominal metres of each overburden layer above competent rock;
// `rockThickness` is the competent-rock band above deep bedrock. Numbers are
// hand-tuned to read like each city's ground, not survey data.

interface GeologyProfile {
  /** nominal fill/soil thickness (m) */
  soil: number;
  /** nominal clay/sand overburden thickness (m) */
  clay: number;
  /** nominal competent-rock band thickness before bedrock (m) */
  rockThickness: number;
  /** 0..1 hardness of the competent rock (drives bored-TBM cost) */
  rockHardness: number;
  /** nominal water-table depth (m) on flat ground at sea level */
  baseWaterTable: number;
  /** 0..1 wetness: higher pulls the table up (low-lying, near water) */
  wetness: number;
  /** metres the water table drops per metre of surface elevation */
  wtElevFactor: number;
}

const GENERIC_PROFILE: GeologyProfile = {
  soil: 4, clay: 15, rockThickness: 40, rockHardness: 0.55,
  baseWaterTable: 8, wetness: 0.35, wtElevFactor: 0.4,
};

const CITY_PROFILES: Record<string, GeologyProfile> = {
  generic: GENERIC_PROFILE,
  // Manhattan schist: thin overburden, competent rock at ~10-20 m, very hard.
  nyc: { soil: 4, clay: 8, rockThickness: 60, rockHardness: 0.88, baseWaterTable: 8, wetness: 0.4, wtElevFactor: 0.45 },
  // Boston: deep made-ground fill over thick soft Boston Blue Clay, rock deep,
  // shallow table on the made land near the harbour.
  boston: { soil: 8, clay: 22, rockThickness: 45, rockHardness: 0.5, baseWaterTable: 4, wetness: 0.6, wtElevFactor: 0.35 },
  // Chicago: very deep soft glacial clay over dolomite; classic soft-ground bore.
  chicago: { soil: 3, clay: 28, rockThickness: 40, rockHardness: 0.45, baseWaterTable: 5, wetness: 0.45, wtElevFactor: 0.3 },
  // San Francisco: mixed ground plus soft Bay Mud in the low made land.
  sf: { soil: 5, clay: 14, rockThickness: 50, rockHardness: 0.62, baseWaterTable: 4, wetness: 0.55, wtElevFactor: 0.5 },
  // Seattle: dense glacial till (hard to bore but not rock), rock deep.
  seattle: { soil: 4, clay: 26, rockThickness: 45, rockHardness: 0.55, baseWaterTable: 6, wetness: 0.45, wtElevFactor: 0.45 },
  // Los Angeles: deep soft alluvium, rock very deep, deep table (semi-arid).
  la: { soil: 6, clay: 42, rockThickness: 35, rockHardness: 0.3, baseWaterTable: 12, wetness: 0.2, wtElevFactor: 0.35 },
  // Washington DC: Atlantic coastal-plain sands/clays over deeper rock.
  dc: { soil: 5, clay: 24, rockThickness: 40, rockHardness: 0.42, baseWaterTable: 5, wetness: 0.5, wtElevFactor: 0.4 },
  // Philadelphia: Wissahickon schist under a modest coastal-plain wedge.
  philly: { soil: 5, clay: 15, rockThickness: 50, rockHardness: 0.62, baseWaterTable: 6, wetness: 0.4, wtElevFactor: 0.45 },
  // Atlanta: residual saprolite soil over hard Piedmont granite, upland table.
  atlanta: { soil: 6, clay: 15, rockThickness: 55, rockHardness: 0.78, baseWaterTable: 10, wetness: 0.25, wtElevFactor: 0.5 },
  // Cleveland: glacial clay over shale, cloudy-wet lowland table.
  cleveland: { soil: 4, clay: 20, rockThickness: 40, rockHardness: 0.5, baseWaterTable: 5, wetness: 0.45, wtElevFactor: 0.35 },
};

/** Resolve a city's geology profile, falling back to the generic temperate one. */
export function geologyProfile(cityKey: string | undefined): GeologyProfile {
  return (cityKey && CITY_PROFILES[cityKey]) || GENERIC_PROFILE;
}

// ── Seeded value-noise (deterministic, no RNG-stream draw) ───────────────────
// Independent, fixed salts so geology never collides with the main RNG stream
// or with weather. Separate band/table salts keep the two fields uncorrelated.
const BAND_SALT = 0x51ed270b;
const TABLE_SALT = 0x2c1b3c6d;

/** Cell size (m) of the coarse noise lattice; band depths vary smoothly on this. */
export const STRATA_NOISE_CELL = 250;
/** ±fraction a layer thickness can wander from nominal. */
export const STRATA_NOISE_FRAC = 0.35;

function hash2(seed: number, salt: number, ix: number, iy: number): number {
  let h = (seed ^ salt) >>> 0;
  h = Math.imul(h ^ (ix + 0x9e3779b9), 0x85ebca6b) >>> 0;
  h = Math.imul(h ^ (iy + 0x165667b1), 0xc2b2ae35) >>> 0;
  h = (h ^ (h >>> 15)) >>> 0;
  return h / 4294967296; // [0,1)
}

/** Smooth bilinear value noise in [0,1) at world (x,y) on the coarse lattice. */
function valueNoise(seed: number, salt: number, x: number, y: number): number {
  const gx = x / STRATA_NOISE_CELL;
  const gy = y / STRATA_NOISE_CELL;
  const ix = Math.floor(gx);
  const iy = Math.floor(gy);
  const fx = gx - ix;
  const fy = gy - iy;
  // smoothstep for continuity
  const sx = fx * fx * (3 - 2 * fx);
  const sy = fy * fy * (3 - 2 * fy);
  const v00 = hash2(seed, salt, ix, iy);
  const v10 = hash2(seed, salt, ix + 1, iy);
  const v01 = hash2(seed, salt, ix, iy + 1);
  const v11 = hash2(seed, salt, ix + 1, iy + 1);
  return (v00 * (1 - sx) + v10 * sx) * (1 - sy) + (v01 * (1 - sx) + v11 * sx) * sy;
}

/** Signed ±STRATA_NOISE_FRAC multiplier for a layer, from a salted noise field. */
function thicknessJitter(seed: number, salt: number, x: number, y: number): number {
  return 1 + (valueNoise(seed, salt, x, y) - 0.5) * 2 * STRATA_NOISE_FRAC;
}

// ── Elevation sampling (StaticElevation grid, co-registered with the masks) ──
/** Sample the surface elevation (m above sea level) at world (x,y). Mirrors the
 *  mask sampling in osmCity.maskAt so elevation and water line up. Returns 0
 *  (sea level, flat) when a city has no baked elevation (procedural cities). */
export function sampleSurfaceElevation(
  elev: Int16Array | undefined,
  res: number | undefined,
  worldSize: number,
  x: number,
  y: number,
): number {
  if (!elev || !res) return 0;
  const half = worldSize / 2;
  let c = Math.floor(((x + half) / worldSize) * res);
  let r = Math.floor(((y + half) / worldSize) * res);
  if (c < 0) c = 0; else if (c >= res) c = res - 1;
  if (r < 0) r = 0; else if (r >= res) r = res - 1;
  return elev[r * res + c] as number;
}

// ── Water table ──────────────────────────────────────────────────────────────
/** Water table can never be modelled shallower/deeper than these bounds. */
export const MIN_WATER_TABLE = 1.5;
export const MAX_WATER_TABLE = 45;

/**
 * Depth (m) to the water table at a column. Low-lying ground near water level
 * (small surface elevation) has a SHALLOW table; high ground has a deep one.
 * `surfaceElevation` is taken as the height above the nearest water surface,
 * approximated as sea level (0) — the case that matters for the coastal /
 * riverine cities where tunnelling meets groundwater.
 */
export function waterTableDepthAt(profile: GeologyProfile, surfaceElevation: number, seed: number, x: number, y: number): number {
  const noise = (valueNoise(seed, TABLE_SALT, x, y) - 0.5) * 2 * 3; // ±3 m
  const wt =
    profile.baseWaterTable * (1 - 0.4 * profile.wetness) +
    Math.max(0, surfaceElevation) * profile.wtElevFactor +
    noise;
  if (wt < MIN_WATER_TABLE) return MIN_WATER_TABLE;
  if (wt > MAX_WATER_TABLE) return MAX_WATER_TABLE;
  return wt;
}

// ── Column reconstruction ────────────────────────────────────────────────────
/**
 * Reconstruct the full subsurface column at world (x,y) for a city. Pure
 * function of (seed, profile, elevation) — no RNG-stream draw, O(1).
 */
export function strataColumn(
  profile: GeologyProfile,
  seed: number,
  worldSize: number,
  elev: Int16Array | undefined,
  elevRes: number | undefined,
  p: Vec2,
): StrataColumn {
  const surfaceElevation = sampleSurfaceElevation(elev, elevRes, worldSize, p.x, p.y);
  // three independent salted noise draws for the three layer boundaries
  const soilBottom = profile.soil * thicknessJitter(seed, BAND_SALT, p.x, p.y);
  const clayBottom = soilBottom + profile.clay * thicknessJitter(seed, BAND_SALT ^ 0x1111, p.x, p.y);
  const bedrockTop = clayBottom + profile.rockThickness * thicknessJitter(seed, BAND_SALT ^ 0x2222, p.x, p.y);
  const bands: StrataBand[] = [
    { kind: 'fill', top: 0, bottom: soilBottom },
    { kind: 'clay', top: soilBottom, bottom: clayBottom },
    { kind: 'rock', top: clayBottom, bottom: bedrockTop },
    { kind: 'bedrock', top: bedrockTop, bottom: BEDROCK_NOMINAL_BOTTOM },
  ];
  return {
    bands,
    waterTableDepth: waterTableDepthAt(profile, surfaceElevation, seed, p.x, p.y),
    rockHardness: profile.rockHardness,
    surfaceElevation,
  };
}

/** Which stratum sits at a given depth (m below surface) in a column. */
export function stratumAtDepth(col: StrataColumn, depth: number): Stratum {
  for (const b of col.bands) {
    if (depth < b.bottom) return b.kind;
  }
  return 'bedrock';
}

/** Convenience: build a column straight from GameState-shaped inputs. */
export function columnAt(
  cityKey: string | undefined,
  seed: number,
  worldSize: number,
  elev: Int16Array | undefined,
  elevRes: number | undefined,
  p: Vec2,
): StrataColumn {
  return strataColumn(geologyProfile(cityKey), seed, worldSize, elev, elevRes, p);
}

File: crates/mf-sim/src/geology.rs
//! Strata / geology model (v0.8 Underground -- sim side).
//!
//! Port of `sim/src/core/geology.ts`. The subsurface at any column `(x, y)` is
//! a pure function of `(seed, x, y, city geology profile)`: there is NO stored
//! 3D grid, band depths are reconstructed on demand from seeded value-noise, so
//! the model costs O(1) memory and reproduces bit-for-bit. It draws only from a
//! DERIVED salted noise field (never `state.rng_state`), so enabling geology
//! cannot perturb the existing city-event / growth RNG stream.
//!
//! The tunable *economics* live in the sibling module [`crate::geology_cost`].

use crate::geometry::Vec2;

/// Top-down stratum kinds. `Fill` = made ground / soil, `Clay` = clay/sand
/// mixed overburden, `Rock` = competent rock, `Bedrock` = deep basement rock.
#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub enum Stratum {
    /// Made ground / soil.
    Fill,
    /// Clay / sand mixed overburden.
    Clay,
    /// Competent rock.
    Rock,
    /// Deep basement rock.
    Bedrock,
}

impl Stratum {
    /// Lowercase string name (mirrors the TS union member).
    pub fn as_str(self) -> &'static str {
        match self {
            Stratum::Fill => "fill",
            Stratum::Clay => "clay",
            Stratum::Rock => "rock",
            Stratum::Bedrock => "bedrock",
        }
    }
}

/// Canonical top-down order.
pub const STRATA: [Stratum; 4] = [
    Stratum::Fill,
    Stratum::Clay,
    Stratum::Rock,
    Stratum::Bedrock,
];

/// A single band in a reconstructed column. Depths are metres BELOW the surface
/// (`top < bottom`); the bedrock band uses a large nominal bottom.
#[derive(Clone, Copy, Debug, PartialEq)]
pub struct StrataBand {
    /// Stratum kind of this band.
    pub kind: Stratum,
    /// Depth (m) to the top of the band.
    pub top: f64,
    /// Depth (m) to the bottom of the band.
    pub bottom: f64,
}

/// A fully reconstructed subsurface column at one `(x, y)`.
#[derive(Clone, Debug, PartialEq)]
pub struct StrataColumn {
    /// Top-down bands.
    pub bands: [StrataBand; 4],
    /// Depth (m below surface) to the top of the water table.
    pub water_table_depth: f64,
    /// 0..1 -- how hard the competent rock is to cut (schist/granite high).
    pub rock_hardness: f64,
    /// Surface elevation (m above sea level) used to build this column.
    pub surface_elevation: f64,
}

/// Nominal bottom of the (unbounded) bedrock band, metres.
pub const BEDROCK_NOMINAL_BOTTOM: f64 = 1000.0;

/// Per-city geology profile (content, like the climate tables).
#[derive(Clone, Copy, Debug, PartialEq)]
pub struct GeologyProfile {
    /// Nominal fill/soil thickness (m).
    pub soil: f64,
    /// Nominal clay/sand overburden thickness (m).
    pub clay: f64,
    /// Nominal competent-rock band thickness before bedrock (m).
    pub rock_thickness: f64,
    /// 0..1 hardness of the competent rock (drives bored-TBM cost).
    pub rock_hardness: f64,
    /// Nominal water-table depth (m) on flat ground at sea level.
    pub base_water_table: f64,
    /// 0..1 wetness: higher pulls the table up (low-lying, near water).
    pub wetness: f64,
    /// Metres the water table drops per metre of surface elevation.
    pub wt_elev_factor: f64,
}

/// Temperate baseline; every unknown city falls back to this.
pub const GENERIC_PROFILE: GeologyProfile = GeologyProfile {
    soil: 4.0,
    clay: 15.0,
    rock_thickness: 40.0,
    rock_hardness: 0.55,
    base_water_table: 8.0,
    wetness: 0.35,
    wt_elev_factor: 0.4,
};

/// Resolve a city's geology profile, falling back to the generic temperate one.
pub fn geology_profile(city_key: Option<&str>) -> GeologyProfile {
    match city_key {
        Some("generic") => GENERIC_PROFILE,
        Some("nyc") => GeologyProfile {
            soil: 4.0,
            clay: 8.0,
            rock_thickness: 60.0,
            rock_hardness: 0.88,
            base_water_table: 8.0,
            wetness: 0.4,
            wt_elev_factor: 0.45,
        },
        Some("boston") => GeologyProfile {
            soil: 8.0,
            clay: 22.0,
            rock_thickness: 45.0,
            rock_hardness: 0.5,
            base_water_table: 4.0,
            wetness: 0.6,
            wt_elev_factor: 0.35,
        },
        Some("chicago") => GeologyProfile {
            soil: 3.0,
            clay: 28.0,
            rock_thickness: 40.0,
            rock_hardness: 0.45,
            base_water_table: 5.0,
            wetness: 0.45,
            wt_elev_factor: 0.3,
        },
        Some("sf") => GeologyProfile {
            soil: 5.0,
            clay: 14.0,
            rock_thickness: 50.0,
            rock_hardness: 0.62,
            base_water_table: 4.0,
            wetness: 0.55,
            wt_elev_factor: 0.5,
        },
        Some("seattle") => GeologyProfile {
            soil: 4.0,
            clay: 26.0,
            rock_thickness: 45.0,
            rock_hardness: 0.55,
            base_water_table: 6.0,
            wetness: 0.45,
            wt_elev_factor: 0.45,
        },
        Some("la") => GeologyProfile {
            soil: 6.0,
            clay: 42.0,
            rock_thickness: 35.0,
            rock_hardness: 0.3,
            base_water_table: 12.0,
            wetness: 0.2,
            wt_elev_factor: 0.35,
        },
        Some("dc") => GeologyProfile {
            soil: 5.0,
            clay: 24.0,
            rock_thickness: 40.0,
            rock_hardness: 0.42,
            base_water_table: 5.0,
            wetness: 0.5,
            wt_elev_factor: 0.4,
        },
        Some("philly") => GeologyProfile {
            soil: 5.0,
            clay: 15.0,
            rock_thickness: 50.0,
            rock_hardness: 0.62,
            base_water_table: 6.0,
            wetness: 0.4,
            wt_elev_factor: 0.45,
        },
        Some("atlanta") => GeologyProfile {
            soil: 6.0,
            clay: 15.0,
            rock_thickness: 55.0,
            rock_hardness: 0.78,
            base_water_table: 10.0,
            wetness: 0.25,
            wt_elev_factor: 0.5,
        },
        Some("cleveland") => GeologyProfile {
            soil: 4.0,
            clay: 20.0,
            rock_thickness: 40.0,
            rock_hardness: 0.5,
            base_water_table: 5.0,
            wetness: 0.45,
            wt_elev_factor: 0.35,
        },
        _ => GENERIC_PROFILE,
    }
}

// -- Seeded value-noise (deterministic, no RNG-stream draw) -------------------
const BAND_SALT: u32 = 0x51ed_270b;
const TABLE_SALT: u32 = 0x2c1b_3c6d;

/// Cell size (m) of the coarse noise lattice; band depths vary smoothly on this.
pub const STRATA_NOISE_CELL: f64 = 250.0;
/// +/- fraction a layer thickness can wander from nominal.
pub const STRATA_NOISE_FRAC: f64 = 0.35;

/// Integer hash to `[0,1)`. Mirrors the JS `hash2` bit-for-bit: every op is a
/// low-32-bit integer operation, identical whether read signed or unsigned.
fn hash2(seed: u32, salt: u32, ix: i32, iy: i32) -> f64 {
    let mut h = seed ^ salt;
    let t1 = (ix as u32).wrapping_add(0x9e37_79b9);
    h = (h ^ t1).wrapping_mul(0x85eb_ca6b);
    let t2 = (iy as u32).wrapping_add(0x1656_67b1);
    h = (h ^ t2).wrapping_mul(0xc2b2_ae35);
    h ^= h >> 15;
    f64::from(h) / 4_294_967_296.0
}

/// Smooth bilinear value noise in `[0,1)` at world `(x,y)` on the coarse lattice.
fn value_noise(seed: u32, salt: u32, x: f64, y: f64) -> f64 {
    let gx = x / STRATA_NOISE_CELL;
    let gy = y / STRATA_NOISE_CELL;
    let ix = gx.floor() as i32;
    let iy = gy.floor() as i32;
    let fx = gx - f64::from(ix);
    let fy = gy - f64::from(iy);
    let sx = fx * fx * (3.0 - 2.0 * fx);
    let sy = fy * fy * (3.0 - 2.0 * fy);
    let v00 = hash2(seed, salt, ix, iy);
    let v10 = hash2(seed, salt, ix + 1, iy);
    let v01 = hash2(seed, salt, ix, iy + 1);
    let v11 = hash2(seed, salt, ix + 1, iy + 1);
    (v00 * (1.0 - sx) + v10 * sx) * (1.0 - sy) + (v01 * (1.0 - sx) + v11 * sx) * sy
}

/// Signed +/-`STRATA_NOISE_FRAC` multiplier for a layer.
fn thickness_jitter(seed: u32, salt: u32, x: f64, y: f64) -> f64 {
    1.0 + (value_noise(seed, salt, x, y) - 0.5) * 2.0 * STRATA_NOISE_FRAC
}

/// Sample the surface elevation (m above sea level) at world `(x,y)`. Returns 0
/// (sea level, flat) when a city has no baked elevation (procedural cities).
pub fn sample_surface_elevation(
    elev: Option<&[i16]>,
    res: Option<u32>,
    world_size: f64,
    x: f64,
    y: f64,
) -> f64 {
    let (elev, res) = match (elev, res) {
        (Some(e), Some(r)) if r > 0 => (e, r as i64),
        _ => return 0.0,
    };
    let half = world_size / 2.0;
    let mut c = (((x + half) / world_size) * res as f64).floor() as i64;
    let mut r = (((y + half) / world_size) * res as f64).floor() as i64;
    if c < 0 {
        c = 0;
    } else if c >= res {
        c = res - 1;
    }
    if r < 0 {
        r = 0;
    } else if r >= res {
        r = res - 1;
    }
    f64::from(elev[(r * res + c) as usize])
}

// -- Water table --------------------------------------------------------------
/// Minimum modelled water-table depth (m).
pub const MIN_WATER_TABLE: f64 = 1.5;
/// Maximum modelled water-table depth (m).
pub const MAX_WATER_TABLE: f64 = 45.0;

/// Depth (m) to the water table at a column.
pub fn water_table_depth_at(
    profile: &GeologyProfile,
    surface_elevation: f64,
    seed: u32,
    x: f64,
    y: f64,
) -> f64 {
    let noise = (value_noise(seed, TABLE_SALT, x, y) - 0.5) * 2.0 * 3.0; // +/-3 m
    let wt = profile.base_water_table * (1.0 - 0.4 * profile.wetness)
        + surface_elevation.max(0.0) * profile.wt_elev_factor
        + noise;
    wt.clamp(MIN_WATER_TABLE, MAX_WATER_TABLE)
}

// -- Column reconstruction ----------------------------------------------------
/// Reconstruct the full subsurface column at world `p` for a city. Pure
/// function of `(seed, profile, elevation)` -- no RNG-stream draw, O(1).
pub fn strata_column(
    profile: &GeologyProfile,
    seed: u32,
    world_size: f64,
    elev: Option<&[i16]>,
    elev_res: Option<u32>,
    p: Vec2,
) -> StrataColumn {
    let surface_elevation = sample_surface_elevation(elev, elev_res, world_size, p.x, p.y);
    let soil_bottom = profile.soil * thickness_jitter(seed, BAND_SALT, p.x, p.y);
    let clay_bottom =
        soil_bottom + profile.clay * thickness_jitter(seed, BAND_SALT ^ 0x1111, p.x, p.y);
    let bedrock_top =
        clay_bottom + profile.rock_thickness * thickness_jitter(seed, BAND_SALT ^ 0x2222, p.x, p.y);
    let bands = [
        StrataBand {
            kind: Stratum::Fill,
            top: 0.0,
            bottom: soil_bottom,
        },
        StrataBand {
            kind: Stratum::Clay,
            top: soil_bottom,
            bottom: clay_bottom,
        },
        StrataBand {
            kind: Stratum::Rock,
            top: clay_bottom,
            bottom: bedrock_top,
        },
        StrataBand {
            kind: Stratum::Bedrock,
            top: bedrock_top,
            bottom: BEDROCK_NOMINAL_BOTTOM,
        },
    ];
    StrataColumn {
        bands,
        water_table_depth: water_table_depth_at(profile, surface_elevation, seed, p.x, p.y),
        rock_hardness: profile.rock_hardness,
        surface_elevation,
    }
}

/// Which stratum sits at a given depth (m below surface) in a column.
pub fn stratum_at_depth(col: &StrataColumn, depth: f64) -> Stratum {
    for b in &col.bands {
        if depth < b.bottom {
            return b.kind;
        }
    }
    Stratum::Bedrock
}

/// Convenience: build a column straight from GameState-shaped inputs.
#[allow(clippy::too_many_arguments)]
pub fn column_at(
    city_key: Option<&str>,
    seed: u32,
    world_size: f64,
    elev: Option<&[i16]>,
    elev_res: Option<u32>,
    p: Vec2,
) -> StrataColumn {
    strata_column(
        &geology_profile(city_key),
        seed,
        world_size,
        elev,
        elev_res,
        p,
    )
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::geometry::vec;

    #[test]
    fn column_is_deterministic_run_twice() {
        let p = vec(1234.5, -678.9);
        let a = column_at(Some("nyc"), 12345, 12000.0, None, None, p);
        let b = column_at(Some("nyc"), 12345, 12000.0, None, None, p);
        assert_eq!(a, b);
    }

    #[test]
    fn bands_are_ordered_and_contiguous() {
        let col = column_at(Some("boston"), 999, 12000.0, None, None, vec(500.0, 500.0));
        assert_eq!(col.bands[0].top, 0.0);
        for w in col.bands.windows(2) {
            assert!((w[0].bottom - w[1].top).abs() < 1e-9);
            assert!(w[0].bottom <= w[1].bottom);
        }
        assert_eq!(col.bands[3].bottom, BEDROCK_NOMINAL_BOTTOM);
    }

    #[test]
    fn water_table_within_bounds() {
        for city in ["nyc", "boston", "chicago", "sf", "la", "generic"] {
            let col = column_at(Some(city), 7, 12000.0, None, None, vec(-2000.0, 3000.0));
            assert!(col.water_table_depth >= MIN_WATER_TABLE);
            assert!(col.water_table_depth <= MAX_WATER_TABLE);
        }
    }

    #[test]
    fn stratum_at_depth_walks_bands() {
        let col = column_at(Some("generic"), 42, 12000.0, None, None, vec(0.0, 0.0));
        assert_eq!(stratum_at_depth(&col, 0.0), Stratum::Fill);
        assert_eq!(stratum_at_depth(&col, 2000.0), Stratum::Bedrock);
    }
}
